src/extension.ts

```diff
@@ -108,34 +108,52 @@ export function activate(context: ExtensionContext) {
     );
 }
 
+async function exec(command: string, args: string[]): Promise<{ stdout: string, stderr: string }> {
+    return new Promise<{ stdout: string, stderr: string }>((resolve, reject) => {
+        let out = '';
+        let err = '';
+        const cp = spawn(command, args);
+        cp.stdout.on('data', data => out += data.toString());
+        cp.stderr.on('data', data => err += data.toString());
+        cp.on('error', reject);
+        cp.on('close', (code, signal) => {
+            const stdout = out.trim();
+            const stderr = err.trim();
+            if (code) {
+                reject(new Error(stderr ?? `Exit code: ${code}`));
+            }
+            if (signal) {
+                reject(new Error(stderr ?? `Exit signal: ${signal}`));
+            }
+            if (err.trim()) {
+                reject(new Error(stderr));
+            }
+            resolve({ stdout, stderr });
+        });
+    });
+}
+
 async function startDebug(_: ExtensionContext, config: DebugConfig): Promise<boolean> {
     let info: DebugInfo | undefined = undefined;
-    let rawStdout: string | undefined = undefined;
-    let rawStdErr: string | undefined = undefined;
+    let stderr: string | undefined = undefined;
+    let stdout: string | undefined = undefined;
     try {
         const args = ['debug', '-I', '-b', config.board.fqbn, config.sketchPath, '--format', 'json'];
-        const { stdout, stderr } = spawnSync(config?.cliPath || '.', args, { encoding: 'utf8' });
-        rawStdout = stdout.trim();
-        rawStdErr = stderr.trim();
+        const result = await exec(config?.cliPath || '.', args);
+        stdout = result.stdout;
+        stderr = result.stderr;
     } catch (err) {
-        showError(err);
-        return false;
+        throw err;
     }
-    if (!rawStdout) {
-        if (rawStdErr) {
-            if (rawStdErr.toLowerCase().indexOf('compiled sketch not found in') !== -1) {
-                vscode.window.showErrorMessage(`Sketch '${path.basename(config.sketchPath)}' was not compiled. Please compile the sketch and start debugging again.`);
-            } else {
-                vscode.window.showErrorMessage(rawStdErr);
-            }
+    if (!stdout) {
+        if (stderr) {
+            throw new Error(stderr);
         }
-        return false;
     }
     try {
-        info = JSON.parse(rawStdout);
+        info = JSON.parse(stdout);
     } catch (err) {
-        console.error(`Could not parse JSON: <${rawStdout}>`);
-        showError(err);
+        throw err;
     }
     if (!info) {
         return false;
```

Comment:
This code looks fine to me, but I would just like to suggest a little rearranging here to make it a bit more readable:
```ts
let info: DebugInfo | undefined = undefined;
    try {
        const args = ['debug', '-I', '-b', config.board.fqbn, config.sketchPath, '--format', 'json'];
        const { stdout, stderr } = await exec(config?.cliPath || '.', args);
        if (!stdout && stderr) {
            throw new Error(stderr);
        }
        info = JSON.parse(stdout);
        if (!info) {
            return false;
        }
    } catch (err) {
        throw err;
    }
```

I moved `stdout` and `stderr` inside the first `try{} catch{}`. That way we can avoid having another `try{} catch{}` block.

@kittaakos what do you think?

Reply:
Good idea. Thank you!

Reply:
7ef1b58